### PR TITLE
[DSW-200] Allow SSL in server's internal SMTP mailer

### PR DESCRIPTION
**Pull request by MarekSuchanek**
- Improved the internal mailer to allow SSL and custom ports.
- Optional configuration options `port` and `usessl` added (needs to be included in the documentation when deployed). 
- Email templates improved (using configurable sender's name).

### CODE REVIEW

File: config/app-config.cfg.example [hunk 1]
```diff
@@ -39,6 +39,8 @@ enabled = false
 name =
 email =
 host =
+port =
+usessl =
```
Comment (vknaisl) [inline config/app-config.cfg.example]: I would name it just `ssl` for simplification

File: lib/Model/Config/AppConfig.hs [hunk 2]
```diff
@@ -50,6 +50,8 @@ data AppConfigMail = AppConfigMail
   , _appConfigMailName :: String
   , _appConfigMailEmail :: String
   , _appConfigMailHost :: String
+  , _appConfigMailPort :: Maybe Int
+  , _appConfigMailUseSSL :: Bool
```
Comment (vknaisl) [inline lib/Model/Config/AppConfig.hs]: Same as the comment above

**Comment by vknaisl** [comment]
A name of the Pull Request same as a name of the commits should include a number of the development subtask (not the story)

**Review by vknaisl — CHANGES_REQUESTED** [review]

File: config/app-config.cfg.example (hunk 1 above)
Comment (MarekSuchanek) [reply config/app-config.cfg.example]: Solved in https://github.com/ds-wizard/dsw-server/pull/42/commits/3b85b9807c60716a599351ead0ff5d672863e81e

File: lib/Model/Config/AppConfig.hs (hunk 2 above)
Comment (MarekSuchanek) [reply lib/Model/Config/AppConfig.hs]: Solved in https://github.com/ds-wizard/dsw-server/pull/42/commits/3b85b9807c60716a599351ead0ff5d672863e81e

File: lib/Service/Mail/Mailer.hs [hunk 3]
```diff
@@ -6,58 +6,89 @@ module Service.Mail.Mailer
...
+      mailPassword = mailConfig ^. password
+      mailSubject = TL.toStrict subject
+      mailMessage = createEmail addrFrom addrTo mailSubject plainBody htmlBody []
+  if mailConfig ^. enabled
+    then liftIO $ makeConnection mailSSL mailHost mailPort $ \connection -> do
+           authSuccess <- SMTP.authenticate Auth.LOGIN mailUsername mailPassword connection
+           renderedMail <- MIME.renderMail' mailMessage
+           if authSuccess
+             then SMTP.sendMail from [to] (S.concat . B.toChunks $ renderedMail) connection
+             else return ()
```
Comment (vknaisl) [inline lib/Service/Mail/Mailer.hs]: We should log when email was sent (probably just as an INFO) and when it failed (I would log it as a WARNING)

File: lib/Service/Mail/Mailer.hs [hunk 4]
```diff
@@ -6,58 +6,89 @@ module Service.Mail.Mailer
...
-  createEmail email subject body
+      mailName = dswConfig ^. mail . name
+      subject = TL.pack $ mailName ++ ": Reset Password"
+      body = TL.pack $ "Hi! You can set up a new password " ++ link ++ "! " ++ mailName ++ " Team"
+  sendEmail email subject body
+
+-- --------------------------------
+-- PRIVATE
+-- --------------------------------
+createEmail to from subject plainBody "" _ = MIME.simpleMail' to from subject plainBody
```
Comment (vknaisl) [inline lib/Service/Mail/Mailer.hs]: If it is not so hard, I would add here type specification

File: lib/Service/Mail/Mailer.hs [hunk 5]
```diff
@@ -6,58 +6,89 @@ module Service.Mail.Mailer
...
+createEmail to from subject plainBody htmlBody attachments =
+  MIME.simpleMailInMemory to from subject plainBody htmlBody attachments
+
+makeConnection False host Nothing = SMTP.doSMTP host
```
Comment (vknaisl) [inline lib/Service/Mail/Mailer.hs:66]: If it is not so hard, I would add here type specification

**Review by vknaisl — CHANGES_REQUESTED** [review]

File: lib/Service/Mail/Mailer.hs (hunk 5 above)
Comment (MarekSuchanek) [reply lib/Service/Mail/Mailer.hs:66]: As I remember, I think here is a problem with type specification that I would have to create own typeclass for that... I will try.

Comment (MarekSuchanek) [reply lib/Service/Mail/Mailer.hs:66]: Done in https://github.com/ds-wizard/dsw-server/pull/42/commits/69bc28099e3a562cff14378da6cb796e4b687ce6

File: lib/Service/Mail/Mailer.hs (hunk 4 above)
Comment (MarekSuchanek) [reply lib/Service/Mail/Mailer.hs]: Done in https://github.com/ds-wizard/dsw-server/pull/42/commits/69bc28099e3a562cff14378da6cb796e4b687ce6 (then this function removed in https://github.com/ds-wizard/dsw-server/pull/42/commits/b29cff1403d1834a815bc90290f3e23ed14c8846)

File: lib/Service/Mail/Mailer.hs (hunk 3 above)
Comment (MarekSuchanek) [reply lib/Service/Mail/Mailer.hs]: Done in https://github.com/ds-wizard/dsw-server/pull/42/commits/cfdd56a7ba1b9269c9d8c7afc343542eec931418 (Do we want email addresses in log?).

It still shows that email has been sent in client, but needs refactoring on side of services (for example - password recovery should be roll-backed if email is not possible to send).

Comment (vknaisl) [reply lib/Service/Mail/Mailer.hs]: - Please, try to get rid off runStdoutLoggingT here
- I would log an email address
- Rollback would be nice but it is out of the scope of this issue. Please create an issue for it in Jira.

Comment (MarekSuchanek) [reply lib/Service/Mail/Mailer.hs]: Done in https://github.com/ds-wizard/dsw-server/pull/42/commits/9c0744920269af18c9534711948820699f4c8541 (other errors, such as errors when connecting to the server, are logged as well with error message from IO exception). 

Rollbacks in [DSW-205](https://ds-wizard.atlassian.net/browse/DSW-205)

**Review by vknaisl — APPROVED** [review]

**Pull request merged**